File: code/C++/143-Reorder List.cpp
/**
Reorder List
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You must do this in-place without altering the nodes' values.

For example,
Given {1,2,3,4}, reorder it to {1,4,2,3}.


 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void reorderList(ListNode *head) {
        
        if(head==NULL)
          return ;
          
          ListNode *p=head;
          
        int len=1;
        while(p->next!=NULL){
            len++;
            p=p->next;
        }
        
        p=head;
        for(int i=0;i<len/2;i++){
            p=p->next;
        }//p指向链表中点
        
        //然后将后半部分反转,最末接到链表中点上
        
        ListNode *curr=p->next;
        ListNode *next;
        p->next=NULL;
        
        while(curr!=NULL){
            next=curr->next;
            curr->next=p;
            p=curr;
            curr=next;
        }
        
        ListNode *h=head;
        
        while(h!=p&&h->next!=p){
            ListNode *tmpH=h->next;
            ListNode *tmpP=p->next;
            
            h->next=p;
            p->next=tmpH;
            h=tmpH;
            p=tmpP;
        }
        
        return ;
        
    }
};

File: code/Python/143-Reorder List.py
# -*- coding: utf-8 -*
"""
@author: Kai Chen
@file: 143-Reorder List.py
@time: 2020/10/20 9:34
@desc:
Given a singly linked list L: L0→L1→…→Ln-1→Ln,
reorder it to: L0→Ln→L1→Ln-1→L2→Ln-2→…

You may not modify the values in the list's nodes, only nodes itself may be changed.

Example 1:

Given 1->2->3->4, reorder it to 1->4->2->3.
Example 2:

Given 1->2->3->4->5, reorder it to 1->5->2->4->3.
"""


# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution(object):
    def reorderList(self, head):
        """
        :type head: ListNode
        :rtype: None Do not return anything, modify head in-place instead.
        """
        if not head:
            return

        mid = self.middleNode(head)
        l1 = head
        l2 = mid.next
        mid.next = None
        l2 = self.reverseList(l2)
        self.mergeList(l1, l2)

    def middleNode(self, head):
        slow = fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    def reverseList(self, head):
        prev, curr = None, head
        while curr:
            nextTemp = curr.next
            curr.next = prev
            prev = curr
            curr = nextTemp
        return prev

    def mergeList(self, l1, l2):
        while l1 and l2:
            l1_tmp = l1.next
            l2_tmp = l2.next

            l1.next = l2
            l1 = l1_tmp

            l2.next = l1
            l2 = l2_tmp
